cmd/juju/caas/remove.go

```diff
@@ -0,0 +1,129 @@
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caas
+
+import (
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/juju/cloud"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/base"
+	cloudapi "github.com/juju/juju/api/cloud"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/jujuclient"
+)
+
+var usageRemoveCAASSummary = `
+Removes a k8s endpoint from Juju.`[1:]
+
+var usageRemoveCAASDetails = `
+Removes the specified k8s cloud from the controller and user-defined
+cloud details (if it is not in use).
+
+Examples:
+    juju remove-k8s myk8scloud
+    
+See also:
+    add-k8s
+`
+
+// Implemented by cloudapi.Client
+type RemoveCloudAPI interface {
+	RemoveCloud(string) error
+	Close() error
+}
+
+// RemoveCAASCommand is the command that allows you to remove a k8s cloud.
+type RemoveCAASCommand struct {
+	modelcmd.ControllerCommandBase
+
+	// cloudName is the name of the caas cloud to remove.
+	cloudName string
+
+	cloudMetadataStore  CloudMetadataStore
+	fileCredentialStore jujuclient.CredentialStore
+	apiRoot             api.Connection
+	removeCloudAPI      func(base.APICallCloser) RemoveCloudAPI
+}
+
+// NewRemoveCAASCommand returns a command to add caas information.
+func NewRemoveCAASCommand(cloudMetadataStore CloudMetadataStore) cmd.Command {
+	cmd := &RemoveCAASCommand{
+		cloudMetadataStore:  cloudMetadataStore,
+		fileCredentialStore: jujuclient.NewFileCredentialStore(),
+		removeCloudAPI: func(caller base.APICallCloser) RemoveCloudAPI {
+			return cloudapi.NewClient(caller)
+		},
+	}
+	return modelcmd.WrapController(cmd)
+}
+
+// Info returns help information about the command.
+func (c *RemoveCAASCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "remove-k8s",
+		Args:    "<k8s name>",
+		Purpose: usageRemoveCAASSummary,
+		Doc:     usageRemoveCAASDetails,
+	}
+}
+
+// SetFlags initializes the flags supported by the command.
+func (c *RemoveCAASCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.CommandBase.SetFlags(f)
+}
+
+// Init populates the command with the args from the command line.
+func (c *RemoveCAASCommand) Init(args []string) (err error) {
+	if len(args) == 0 {
+		return errors.Errorf("missing k8s name.")
+	}
+	c.cloudName = args[0]
+	return cmd.CheckEmpty(args[1:])
+}
+
+func (c *RemoveCAASCommand) newAPIRoot() (api.Connection, error) {
+	if c.apiRoot != nil {
+		return c.apiRoot, nil
+	}
+	return c.NewAPIRoot()
+}
+
+// Run is defined on the Command interface.
+func (c *RemoveCAASCommand) Run(ctxt *cmd.Context) error {
+	api, err := c.newAPIRoot()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer api.Close()
+
+	cloudClient := c.removeCloudAPI(api)
+	if err := cloudClient.RemoveCloud(c.cloudName); err != nil {
+		return errors.Annotatef(err, "cannot remove k8s cloud from controller")
+	}
+
+	if err := removeCloudFromLocal(c.cloudMetadataStore, c.cloudName); err != nil {
+		return errors.Trace(err)
```

Comment:
Might be worthwhile annotating this too to say that is local removal that has failed?

Reply:
yup